tc/core/polyhedral/cuda/memory_promotion_heuristic.cc

```diff
@@ -719,6 +728,50 @@ void promoteToRegistersBelow(MappedScop& mscop, detail::ScheduleTree* scope) {
   }
 }
 
+/*
+ * Promote to registers below "depth" schedule dimensions.  Split bands if
+ * necessary to create promotion scopes.  Do not promote if it would require
+ * splitting the band mapped to threads as we assume only one band can be
+ * mapped.
+ */
+void promoteToRegistersAtDepth(MappedScop& mscop, size_t depth) {
+  using namespace detail;
+
+  auto root = mscop.scop().scheduleRoot();
+
+  // 1. Collect all bands with a member located at the given depth in the
+  // overall schedule.  Make sure this is the last member of the band by
+  // splitting off the subsequent members into a different band.  Ignore bands
+  // mapped to threads if splitting is required as it would break invariant of
+  // a single band being mapped to threads in a subtree.
```

Comment:
Do we still need this invariant after the mapping has taken place?

Reply:
Some promotion heuristics used to need the depth of thread x, which has been replaced by finding the thread-specific marker and taking the last member of the band immediately above that.  Shared memory promotion in particular.  I am considering to swap the order of promotions and do registers before shared memory, at which point the invariant must be preserved.  I left a to do to analyze this issue separately from everything else.

Reply:
On Wed, Jun 20, 2018 at 02:57:49AM -0700, ftynse wrote:
> Some promotion heuristics used to need the depth of thread x, which has been replaced by
 finding the thread-specific marker and taking the last member of the band immediately abo
ve that.

In what way does this depend on a single band being mapped to threads?

skimo

Reply:
It may not require that, indeed. I am not going to spend time and check if it does in this PR.